packages/react-components/lib/reports/multi-level-menu.tsx

```diff
@@ -0,0 +1,103 @@
+import React, { useState } from 'react';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemText from '@material-ui/core/ListItemText';
+import Collapse from '@material-ui/core/Collapse';
+import ExpandLess from '@material-ui/icons/ExpandLess';
+import ExpandMore from '@material-ui/icons/ExpandMore';
+
+interface ListItemBodyProps {
+  icon?: JSX.Element;
+  title: string;
+}
+
+const ListItemBody = (props: ListItemBodyProps): JSX.Element => {
+  return (
+    <>
+      <ListItemIcon>{props.icon}</ListItemIcon>
+      <ListItemText primary={props.title} />
+    </>
+  );
+};
+
+interface MenuItemProps {
+  icon?: JSX.Element;
+  title: string;
+  items?: {
+    title: string;
+    to: string;
+  }[];
+  onClick?: () => void;
+}
+
+const MenuItem = (props: MenuItemProps): JSX.Element => {
+  return (
+    <ListItem button onClick={props.onClick}>
+      <ListItemBody icon={props.icon} title={props.title} />
+    </ListItem>
+  );
+};
+
+export interface ExpandableMultilevelMenuProps {
+  icon?: JSX.Element;
+  title: string;
+  items: MenuItemProps[];
+  onClick?: () => void;
+}
+
+const ExpandableMenuItem = (props: ExpandableMultilevelMenuProps): JSX.Element => {
+  const { items, icon, title } = props;
+  const [open, setOpen] = useState(false);
+
+  const handleClick = () => {
+    setOpen(!open);
+  };
+
+  return (
+    <div>
+      <ListItem button onClick={handleClick}>
+        <ListItemBody icon={icon} title={title} />
+        {open ? <ExpandLess /> : <ExpandMore />}
+      </ListItem>
+      <Collapse in={open} timeout="auto" unmountOnExit>
+        <MultiLevelMenu menuStructure={items} />
+      </Collapse>
+    </div>
+  );
+};
+
+export interface MultilevelMenuProps {
+  menuStructure: (ExpandableMultilevelMenuProps | MenuItemProps)[];
+}
+
+export const MultiLevelMenu = (props: MultilevelMenuProps): React.ReactElement => {
```

Comment:
This re-renders all items when any of them changes, should we memo `ExpandableMenuItem` and `MenuItem` to optimize it? Or it doesn't really matter since they are simple enough that they have minimal performance impact.

Reply:
Some people said that it doesn't matter at this stage of the app. But I can assure you that if in the future we have a troublemaker child component that triggers rerenders every simple re-render of the parents will count. So, I think we should add a memo as you recommend here.

Reply:
Without memo
![image](https://user-images.githubusercontent.com/11761240/112033702-7a2c4980-8b1c-11eb-869e-0300820d2984.png)

With memo
![image](https://user-images.githubusercontent.com/11761240/112033173-ec505e80-8b1b-11eb-81d5-b8970e96a960.png)